File: algorithm/dp/count_all_valid_pickup.cpp



// 1 + 2 + ... n = n(n+1) / 2
// dp[i] how many valid orders there are using i number
// [P1, D1]
// [P2, D2]
// P2: 3 = 3+2+1
// D2: 3

// 6 = 5+4+3+2+1 = 15*6 = 90
// dp[i] = dp[i-1]*(2*n-1+2*n-2+..1)
// n*(n+1) / 2
// n = 2*n-1
// n(2n-1)

#include <vector>

using namespace std;

// Intuition: Whenever there are existing valid sequence, for each new pickup or delivery
// you think about the valid positions to insert them:
// 1. First insert pickups, it should have 2*(n-1)+1 number of options.
// [P1, D1]
// [P2, D2]
// -> [P2, P1, D1], [P1, P2, D1], [P1, D1, P2]

// 2. Insert delivery
// if pick up is inserted at first slot, then delivery have 2 * (n - 1) + 1 options as well.
// but if pick up is inserted at second slot, then delivery have 2 * (n - 1) options, one less slot option, because
// delivery cannot be inserted at first slot, which makes it before pickup order.

// so for each previous valid sequence, we have 2*(n-1)+1 + 2*(n-1) + ... + 1 options to insert it.
// dp[i] = dp[i-1]*(2*n-1+2*n-2+..1)
// 
// Now we need to figure out how to calculate (2*n-1+2*n-2+..1)
// We have a formula 1+2+...+n = n(n+1) / 2
// now n we have is 2*n-1, the formula becomes = (2*n-1)*2*n/2 = n(2n-1)

// Time: O(n), Space: O(N) -> O(1)
int countOrders(int n) {
    vector<long> dp(n + 1, 0);
    dp[1] = 1;

    int mod = 1e9 + 7;

    for (int i = 2; i <= n; i++) {
        dp[i] = dp[i - 1];
        dp[i] *= i;
        dp[i] = dp[i] % mod;
        dp[i] *= (2 * i - 1);
        dp[i] = dp[i] % mod;
    }

    return dp[n];
}

File: algorithm/dp/python/count_all_valid_pickup.py
"""
Count All Valid Pickup and Delivery Options (LeetCode 1359)

Given n orders, each has a pickup (P_i) and a delivery (D_i), count the number
of valid sequences where every P_i appears before its corresponding D_i.

Intuition: When adding the n-th order to a valid sequence of (n-1) orders
(which has 2*(n-1) items), we can:
  1. Insert the pickup at any of 2*(n-1)+1 positions.
  2. Then insert the delivery after the pickup. If pickup goes in slot k
     (1-indexed), delivery has 2*(n-1)+1 - (k-1) options.
Summing over k from 1..2*(n-1)+1 gives 1+2+...+(2n-1) = n*(2n-1).
Therefore: dp[i] = dp[i-1] * i * (2*i - 1).

Approaches:
  1. countOrders: bottom-up DP with 1D array. Time O(n), Space O(n).
"""


# Time: O(n), Space: O(n) -> O(1)
def countOrders(n: int) -> int:
    dp = [0] * (n + 1)
    dp[1] = 1

    mod = 10**9 + 7

    for i in range(2, n + 1):
        dp[i] = dp[i - 1]
        dp[i] *= i
        dp[i] = dp[i] % mod
        dp[i] *= (2 * i - 1)
        dp[i] = dp[i] % mod

    return dp[n]
